src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/AbstractPluginDescriptor.java

```diff
@@ -0,0 +1,102 @@
+package org.broadinstitute.hellbender.tools.walkers.varianteval;
+
+import org.broadinstitute.barclay.argparser.CommandLineException;
+import org.broadinstitute.barclay.argparser.CommandLinePluginDescriptor;
+
+import java.util.*;
+
+abstract public class AbstractPluginDescriptor<T> extends CommandLinePluginDescriptor {
+    protected final Map<String, T> allDiscoveredPlugins = new HashMap<>();
+
+    protected List<String> pluginPackageNames;
+    protected Class<T> pluginBaseClass;
+    protected Class<T> defaultPluginBaseClass;
+
+    public AbstractPluginDescriptor(List<String> pluginPackageNames, Class pluginBaseClass, Class defaultPluginBaseClass) {
+        this.pluginPackageNames = pluginPackageNames;
+        this.pluginBaseClass = pluginBaseClass;
+        this.defaultPluginBaseClass = defaultPluginBaseClass;
+    }
+
+    @Override
+    public List<String> getPackageNames() {
+        return pluginPackageNames;
+    }
+
+    @Override
+    public Class<T> getPluginBaseClass() {
+        return pluginBaseClass;
+    }
+
+    @Override
+    public T createInstanceForPlugin(Class pluggableClass) throws IllegalAccessException, InstantiationException {
+        T ret;
+        final String simpleName = pluggableClass.getSimpleName();
+
+        //TODO: handle default marker class
+        if (allDiscoveredPlugins.containsKey(simpleName)) {
+            throw new IllegalArgumentException(
+                    String.format("A plugin class name collision was detected (%s/%s). " +
+                                    "Simple names of plugin classes must be unique across packages.",
+                            pluggableClass.getName(),
+                            allDiscoveredPlugins.get(simpleName).getClass().getName())
+            );
+        } else {
+            ret = (T) pluggableClass.newInstance();
+        }
+
+        allDiscoveredPlugins.put(simpleName, ret);
+        return ret;
+    }
+
+    @Override
+    public List<T> getDefaultInstances() {
+        List<T> ret = new ArrayList<>();
+        if (defaultPluginBaseClass != null){
+            for (T c : allDiscoveredPlugins.values()){
+                if (defaultPluginBaseClass.isAssignableFrom(c.getClass()))
+                    ret.add(c);
+            }
+        }
+
+        return ret;
+    }
+
+    @Override
+    public List<T> getResolvedInstances() {
+        List<T> ret = new ArrayList<>();
+        allDiscoveredPlugins.forEach((x,y) -> ret.add(y));
+        return ret;
+    }
+
+    @Override
+    public boolean isDependentArgumentAllowed(Class predecessorClass) {
+        return false;
+    }
+
+    @Override
+    public void validateAndResolvePlugins() throws CommandLineException {
+
+    }
+
+    @Override
+    public Set<String> getAllowedValuesForDescriptorHelp(String longArgName) {
+        return null;
```

Comment:
This should be implemented by sub-classes, otherwise the automatic doc generation won't work; shouldn't be the default behavior for the abstract plugin.

Reply:
also yes